test.cc

```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(p.seen_, 3u);
+  EXPECT_EQ("4,5,6", p.results_);
+}
+
+TEST(Sherlock, WaitingProcessorOfIntsReallyWaits) {
+  auto baz_stream = sherlock::Stream<SortOfInt>("baz");
+  thread delayed_publishing_thread([&baz_stream]() {
+    sleep_for(milliseconds(5));
+    baz_stream.Publish(7);
+    baz_stream.Publish(8);
+    baz_stream.Publish(9);
+  });
+  WaitingProcessorOfSortOfInts p;
+  baz_stream.Subscribe(p);  // This call blocks until the processor declared itself done.
+  // Unlike the `Nice*` processor, the `Waiting*` one will wait for three entries to get processed.
+  EXPECT_EQ(p.seen_, 3u);
+  EXPECT_EQ("7,8,9", p.results_);
+  delayed_publishing_thread.join();
+}
+
+// TODO(dkorolev): Support and test `CaughtUp()`: Add a few pre-populated plus a few delayed entries.
+
+struct Point {
+  int x;
+  int y;
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(x), CEREAL_NVP(y));
+  }
+};
+
+// TODO(dkorolev): This class should be moved into `sherlock.h`.
+template <typename T>
+class ServeJSONOverHTTP {
+ public:
+  ServeJSONOverHTTP(Request&& r)
+      : http_request_scope_(std::move(r)), http_response_(http_request_scope_.SendChunkedResponse()) {}
+
+  inline bool Entry(const T& entry) {
+    http_response_(JSON(entry, "point"));  // TODO(dkorolev): WTF do I have to say JSON() here?
+    ++n_;
+    keep_going_ = (n_ < 3);
+    return keep_going_;
+  }
+
+  inline bool TerminationRequest() {
+    return true;  // Run forever (TODO(dkorolev): No, actually, return false here!
+  }
+
+  // TODO(dkorolev): This is the hack to make the test pass.
+  atomic_bool keep_going_;
+
+ private:
+  Request http_request_scope_;  // Need to keep `Request` in scope, for the lifetime of the chunked response.
+  bricks::net::HTTPServerConnection::ChunkedResponseSender http_response_;
+  size_t n_ = 0;  // This is a hack to actually return after 10 entries.
+
+  ServeJSONOverHTTP() = delete;
+  ServeJSONOverHTTP(const ServeJSONOverHTTP&) = delete;
+  void operator=(const ServeJSONOverHTTP&) = delete;
+  ServeJSONOverHTTP(ServeJSONOverHTTP&&) = delete;
+  void operator=(ServeJSONOverHTTP&&) = delete;
+};
+
+TEST(Sherlock, JSONOverHTTP) {
+  auto time_series = sherlock::Stream<Point>("time_series");
+  thread delayed_publishing_thread([&time_series]() {
+    for (int x = 10; x <= 13; ++x) {
+      sleep_for(milliseconds(5));
+      time_series.Publish(Point{x, x * x});
+    }
+  });
+  HTTP(FLAGS_sherlock_http_test_port).Register("/time_series", [&time_series](Request r) {
+    // TODO(dkorolev): Figure out how to best handle ownership of the parameter wrt destruction.
+    // For now it's a hack.
+    ServeJSONOverHTTP<Point> server(std::move(r));
+    auto tmp = time_series.Subscribe(server);  // TODO(dkorolev): `->Detach()`.
+    while (server.keep_going_) {
+      ;
+    }
+  });
+  EXPECT_EQ("{\"point\":{\"x\":10,\"y\":100}}{\"point\":{\"x\":11,\"y\":121}}{\"point\":{\"x\":12,\"y\":144}}",
```

Comment:
Because I use `JSON()` directly in the serving thread.

It's actually a good thing for now. Once I fix the above one, this test will fail, and I'll fix it here too. :)